usr/iscsiadm.c

```diff
@@ -675,26 +676,122 @@ static int login_portals(struct node_rec *pattern_rec)
 	return rc;
 }
 
-static int print_nodes(int info_level, struct node_rec *rec)
+static void print_node_flat(struct iscsi_node *node)
 {
-	struct node_rec tmp_rec;
+		printf("%s,%" PRIu16 "%s\n",
+		       iscsi_node_portal_get(node),
+		       iscsi_node_tpgt_get(node),
+		       iscsi_node_target_name_get(node));
+}
+
+static void print_nodes_tree(struct iscsi_node **nodes, uint32_t node_count)
+{
+	uint32_t i = 0;
+	struct iscsi_node *cur_node = NULL;
+	struct iscsi_node *pre_node = NULL;
+
+	// According to libopeniscsiusr document, nodes are sorted. There
+	// is no need to create hash table for this.
+	for (i = 0; i < node_count; ++i) {
+		cur_node = nodes[i];
+		if ( i != 0)
+			pre_node = nodes[i - 1];
+		if ((pre_node == NULL) ||
+		    (strcmp(iscsi_node_target_name_get(cur_node),
+			    iscsi_node_target_name_get(pre_node)) != 0))
+			printf("Target: %s\n",
+			       iscsi_node_target_name_get(cur_node));
+		if ((pre_node == NULL) ||
+		    (strcmp(iscsi_node_conn_address_get(cur_node),
+			    iscsi_node_conn_address_get(pre_node)) != 0) ||
+		    (iscsi_node_conn_port_get(cur_node) !=
+		     iscsi_node_conn_port_get(pre_node)))
+			printf("\tPortal: %s,%d\n",
+			       iscsi_node_portal_get(cur_node),
+			       iscsi_node_tpgt_get(cur_node));
+		if ((pre_node == NULL) ||
+		    (strcmp(iscsi_node_iface_name_get(cur_node),
+			    iscsi_node_iface_name_get(pre_node)) != 0))
+			printf("\t\tIface Name: %s\n",
+			       iscsi_node_iface_name_get(cur_node));
+	}
+}
+
+static int print_nodes(struct iscsi_context *ctx, int info_level)
+{
+	struct iscsi_node **nodes = NULL;
+	uint32_t node_count = 0;
+	uint32_t i = 0;
 	int rc = 0;
 
-	switch (info_level) {
-	case 0:
-	case -1:
-		rc = for_each_matched_rec(rec, NULL, idbm_print_node_flat);
-		break;
-	case 1:
-		memset(&tmp_rec, 0, sizeof(node_rec_t));
-		rc = for_each_matched_rec(rec, &tmp_rec,
-					  idbm_print_node_and_iface_tree);
-		break;
-	default:
+	if ((info_level != 0) && (info_level != -1) && (info_level != 1)) {
 		log_error("Invalid info level %d. Try 0 or 1.", info_level);
 		rc = ISCSI_ERR_INVAL;
+		goto out;
```

Comment:
You are jumping to "out" here, but that frees the nodes, which we have not allocatged.

Reply:
`iscsi_nodes_free()` will do nothing against NULL pointer which is already mentioned in API document.